docs/pages/tracing.rst

```diff
@@ -0,0 +1,63 @@
+Tracing
+=======
+
+collect_traces
+--------------
+
+Sometimes we want to trace where the ``Failure`` has occurred in our system,
+`returns` provide a way to trace those failures. By default tracing is disabled.
+
+The trace is accessible by **trace** property that is available for ``Result``,
+``IOResult`` containers. It's basically a list containing all
+:class:`inspect.FrameInfo` from the call stack when the ``Failure`` was created.
+
+To enable it you can use :func:`returns.tracing.collect_traces`.
+See some examples bellow:
+
+You can use it as a context manager:
+
+.. code:: python
+
+  >>> from inspect import FrameInfo
+
+  >>> from returns.result import Failure
+  >>> from returns.tracing import collect_traces
+
+  >>> non_traced_failure = Failure('Normal Failure')
+  >>> with collect_traces():
+  ...     traced_failure = Failure('Traced Failure')
+
+  >>> assert non_traced_failure.trace is None
+  >>> assert isinstance(traced_failure.trace, list)
+  >>> assert isinstance(traced_failure.trace[0], FrameInfo)
+
+Or as a decorator:
+
+.. code:: python
+
+  >>> from inspect import FrameInfo
+
+  >>> from returns.io import IOFailure, IOResult
+  >>> from returns.result import Failure, Result
+  >>> from returns.tracing import collect_traces
+
+  >>> @collect_traces()
+  ... def traced_function(value: str) -> IOResult[str, str]:
+  ...     return IOFailure(value)
+
+  >>> non_traced_failure = Failure('Normal Failure')
+  >>> traced_failure = traced_function('Traced Failure')
+
+  >>> assert non_traced_failure.trace is None
+  >>> assert isinstance(traced_failure.trace, list)
+  >>> assert isinstance(traced_failure.trace[0], FrameInfo)
+
+.. warning::
+
+  Activating trace can make your program a bit slower if it has many points where ``Failure`` is often created.
```

Comment:
We need to add a warning about thread safety. I guess it is not safe to use in multiple threads. It might have side effects. Am I right?

Reply:
I think you're right. But I'll make a test to document this fact!

Reply:
```python
from time import sleep
from typing import Any

from returns.result import Failure
from returns.tracing import collect_traces


def my_func_one(have_to_patch: bool) -> Any:
    if have_to_patch:
        with collect_traces():
            print(f'have_to_patch: {have_to_patch} | sleep')
            sleep(10)
            print(f'Exiting Monkey Patch')
    else:
        print(f'have_to_patch: {have_to_patch} | sleep')
        sleep(5)
        print(f'have_to_patch: {have_to_patch} | trace: {Failure("some value").trace}')
        sleep(6)
        print(f'have_to_patch: {have_to_patch} | trace: {Failure("some value").trace}')


if __name__ == "__main__":
    from concurrent import futures

    with futures.ThreadPoolExecutor(max_workers=2) as executor:
        executor.map(my_func_one, [True, False])
```

**Output**
```text
have_to_patch: True | sleep
have_to_patch: False | sleep
have_to_patch: False | trace: [ OMITTED TRACE ]
Exiting Monkey Patch
have_to_patch: False | trace: None
```

Reply:
I'm going to write the warning now!